File: Editorial/2701-2800/2770_Maximum_Number_of_Jumps_to_Reach_the_Last_Index/TypeScript/1_Memoization_Search.ts
function maximumJumps(nums: number[], target: number): number {
    const n: number = nums.length;
    const memo: number[] = new Array(n).fill(Number.MIN_SAFE_INTEGER);

    const dfs = (i: number): number => {
        if (i === n - 1) {
            return 0;
        }
        if (memo[i] !== Number.MIN_SAFE_INTEGER) {
            return memo[i];
        }

        let res: number = Number.MIN_SAFE_INTEGER;
        for (let j: number = i + 1; j < n; j++) {
            if (Math.abs(nums[i] - nums[j]) <= target) {
                res = Math.max(res, dfs(j) + 1);
            }
        }
        memo[i] = res;
        return res;
    };

    const ans: number = dfs(0);
    return ans < 0 ? -1 : ans;
}

File: Editorial/2701-2800/2770_Maximum_Number_of_Jumps_to_Reach_the_Last_Index/CSharp/1_Memoization_Search.cs
public class Solution {
    public int MaximumJumps(int[] nums, int target) {
        int n = nums.Length;
        int[] memo = new int[n];
        Array.Fill(memo, int.MinValue);

        Func<int, int> dfs = null;
        dfs = (i) => {
            if (i == n - 1) {
                return 0;
            }
            if (memo[i] != int.MinValue) {
                return memo[i];
            }
            int res = int.MinValue;
            for (int j = i + 1; j < n; j++) {
                if (Math.Abs(nums[i] - nums[j]) <= target) {
                    res = Math.Max(res, dfs(j) + 1);
                }
            }
            return memo[i] = res;
        };

        int ans = dfs(0);
        return ans < 0 ? -1 : ans;
    }
}
